Physical units with the pint library

shervin86 opened the pull request:
Several changes:
     - Introduced type checks for values, intervals and options.
     - Using pint.Quantity for floats and physical quantities with units
     - Encode and decode pint.Quantity in json

shervin86 commented on libpyvinyl/Parameters/Collections.py:
```diff
@@ -5,6 +5,38 @@
 
 from libpyvinyl.AbstractBaseClass import AbstractBaseClass
 from .Parameter import Parameter
+from pint.quantity import Quantity
```
Maybe these should be imported from the Parameter?

shervin86 commented on libpyvinyl/Parameters/Collections.py:
```diff
@@ -155,6 +189,10 @@ def to_dict(self):
         for key in self.parameters:
             params[key] = self.parameters[key].__dict__
+            a = params[key]
+            if "_Parameter__value_type" in a:
```
If this is exported, it becomes a mess and it wil be anyway set from the actual values in the from_dict when it is read back.

shervin86 commented on tests/BaseCalculatorTest.py:
```diff
@@ -1,4 +1,5 @@
 import unittest
```
This file has been changed by `black`, so only formatting changes

CFGrote commented on libpyvinyl/Parameters/Collections.py:
```diff
@@ -5,6 +5,38 @@
+from pint.unit import Unit, UnitsContainer
+
+
+def quantity_encode(obj, primitives=False):
```
type hint?

shervin86 replied:
I didn't put any because obj can be anything... so it does not seem to me very informative.
Any suggestion?

CFGrote commented on libpyvinyl/Parameters/Collections.py:
```diff
@@ -5,6 +5,38 @@
 from libpyvinyl.AbstractBaseClass import AbstractBaseClass
 from .Parameter import Parameter
+from pint.quantity import Quantity
+from pint.unit import Unit, UnitsContainer
```
don't forget to add pint to the requirements (if not already there)

shervin86 replied:
It is already there.

CFGrote commented on libpyvinyl/Parameters/Parameter.py:
```diff
@@ -1,34 +1,164 @@
 # Created by Mads Bertelsen and modified by Juncheng E
```
I'd omit these attributions. git knows the authorship of every line in the code

shervin86 replied:
I think we should setup a commented header for each file with authors, the grant, etc.
Git knows but people might just copy the file.
What do you think?

CFGrote replied:
ok

CFGrote reviewed (commented):
I want to play with it a bit, will provide more comments.

CFGrote reviewed (commented):
* Typo: `print_paramter_constraints` -> `print_parameter_constraints`
* Is there a way to get the list of valid (or invalid) intervals? 
* Tests: Class names should be ConvoyCase, not Convoy_Case
* Can we have default values for `min_value=-numpy.inf`, `max_value=numpy.inf`, `intervals_are_legal=True`, or would that break anything?
* I added two tests to ParametersTest.py that shows how I would (try to) use the units system.

shervin86 commented:
> * Typo: `print_paramter_constraints` -> `print_parameter_constraints`
> 
Corrected
>     * Is there a way to get the list of valid (or invalid) intervals?
> 
No, there is no way to get the list back, except from printing.
We can think about it, but I'm not sure about the usage of it. 
The user should not be tempted to manipulate intervals from the eventual list returned by the Parameter class.

>     * Tests: Class names should be ConvoyCase, not Convoy_Case
Done
> 
>     * Can we have default values for `min_value=-numpy.inf`, `max_value=numpy.inf`, `intervals_are_legal=True`, or would that break anything?
It should not break anything. But forcing the user to put both sides of the interval does not seem to be an issue, but rather a good practice.

> 
>     * I added two tests to ParametersTest.py that shows how I would (try to) use the units system.
Ok

JunCEEE commented on libpyvinyl/Parameters/Parameter.py:
```diff
@@ -1,34 +1,163 @@
+        self.__unit = Unit(uni)
+
+    @property
+    def value_no_conversion(self):
```
Two renaming suggestions: `value_pint`, `value_with_unit`.

JunCEEE replied:
Please consider the above suggestion.

JunCEEE commented on libpyvinyl/Parameters/Parameter.py:
```diff
@@ -1,34 +1,247 @@
+         - ...
+        This class method is mainly used to allow dumping and loading the class from json
+        """
+        if not "name" in param_dict:
```
> test for membership should be 'not in'flake8(E713)
Suggestion: 
`if "name" not in param_dict`

JunCEEE commented on libpyvinyl/Parameters/Parameter.py:
```diff
@@ -1,34 +1,247 @@
+    @unit.setter
+    def unit(self, uni):
+        """
+        Assignement of the units
```
Spelling: Assignment of the units

JunCEEE commented on libpyvinyl/Parameters/Parameter.py:
```diff
@@ -1,34 +1,247 @@
+        else:
+            return self.__value
+
+    @classmethod
+    # self is not needed...
+    def __is_type_compatible(self, t1: type, t2: Union[None, type]) -> bool:
```
If `self` is not needed, .e.g. it doesn't need to interact or return the members of the class, I would suggest using `@staticmethod`:
```
@staticmethod
    def __is_type_compatible(t1: type, t2: Union[None, type]) -> bool:
```

JunCEEE commented on libpyvinyl/Parameters/Parameter.py:
```diff
@@ -1,34 +1,247 @@
+            for v in value:
+                if not self.__is_type_compatible(vtype, type(v)):
+                    raise TypeError(
+                        "Iterable object passed as value for the parameter, but it is made of inomogeneous types: ",
```
I guess you mean "inhomogeneous" here?

JunCEEE reviewed (commented):
Generally, I like the `pint` implementation here.
The only thing I would like to stress is that a more intuitive naming for the value in `pint` type would be more friendly to the users.

CFGrote commented:
Looks good to me. Please resolve the pending conflicts and merge.
